core/base/v7/inc/ROOT/TDirectoryEntry.h

```diff
@@ -26,77 +26,70 @@ namespace Experimental {
-  TClass* fType = 0; ///< The type of the object
-
-protected:
-  TDirectoryEntryPtrBase() = default;
-
 public:
-  /// Constructor setting the type.
-  TDirectoryEntryPtrBase(TClass* cl): fType(cl) {}
+  TDirectoryEntry(): TDirectoryEntry(nullptr) {}
+  TDirectoryEntry(std::nullptr_t):
+    TDirectoryEntry(std::make_shared<std::nullptr_t>(nullptr)) {}
+  template<class T>
+  explicit TDirectoryEntry(T* ptr):
+    TDirectoryEntry(std::make_shared<T>(*ptr)) {}
+  template<class T>
+  explicit TDirectoryEntry(const std::shared_ptr<T>& ptr):
+    fTypeInfo(&typeid(T)),
+    fType(TClass::GetClass(typeid(T))),
+    fObj(ptr) {}
 
   /// Get the last change date of the entry.
   const time_point_t& GetDate() const { return fDate; }
 
-  /// Get the object's type.
-  TClass* GetType() const { return fType; }
-
   /// Inform the entry that it has been modified, and needs to update its
   /// last-changed time stamp.
   void SetChanged() { fDate = clock_t::now(); }
 
-  /// Abstract interface to retrieve the address of the object represented by
-  /// this entry, if any.
-  virtual void* GetObjectAddr() const = 0;
-
-  /// Abstract interface to retrieve the type of the object represented by this
-  /// entry.
-  virtual const std::type_info& GetTypeInfo() const = 0;
-};
-
-
-/**
- Type-aware part of an entry in a TDirectory. It can manage the lifetime of an
- object, or only inform about last-change date and the type of an object.
- */
-
-template <class T>
-class TDirectoryEntryPtr: public TDirectoryEntryPtrBase {
-public:
-  /// Initialize a TDirectoryEntryPtr from an existing object ("write").
-  TDirectoryEntryPtr(T&& obj):
-    TDirectoryEntryPtrBase(TClass::GetClass(typeid(T))),  fObj(obj) {}
-
-  /// Initialize a TDirectoryEntryPtr from an existing object ("write").
-  TDirectoryEntryPtr(const std::shared_ptr<T>& ptr): fObj(ptr) {}
+  /// Type of the object represented by this entry.
+  const std::type_info& GetTypeInfo() const { return *fTypeInfo; }
 
-  virtual ~TDirectoryEntryPtr() {}
+  /// Get the object's type.
+  TClass* GetType() const { return fType; }
 
   /// Retrieve the `shared_ptr` of the referenced object.
-  std::shared_ptr<T>& GetPointer() const { return fObj; }
+  std::shared_ptr<void>& GetPointer() { return fObj; }
+  const std::shared_ptr<void>& GetPointer() const { return fObj; }
 
-  /// Retrieve the address of the object managed by this entry.
-  /// Returns `nullptr` if no object is managed by this entry.
-  void* GetObjectAddr() const final { return fObj.get(); }
+  template<class U>
+  std::shared_ptr<U> CastPointer() const;
 
-  /// Type of the object represented by this entry.
-  const std::type_info& GetTypeInfo() const final { return typeid(T); }
+  explicit operator bool() const { return *fTypeInfo != typeid(std::nullptr_t); }
 
 private:
-  std::shared_ptr<T> fObj; ///< Object referenced by this entry, if any.
+  time_point_t fDate = clock_t::now(); ///< Time of last change
```

Comment:
>  This results in each having a different setting for the 'number of spaces' the editor replaces tab with ... resulting in misaligned code.

Solution is surprisingly simple: if project uses tabs for indentation, don't let editor replace tabs with spaces. Symmetrically, one can use editor which doesn't replace tabs with spaces and introduce tabs into space-indented code.
So, what I want to say: the problem is not in tabs.

> In addition, I am not aware on any advantages of using tab characters

You should have started with this! Let me show you some advantages.
First of all, tabs aren't used for anything except indentation, whilst spaces are.
Tabs has "meaning": 1 tab == 1 indent level. If you use e.g. 3 spaces for indent level, what is 2 spaces? ⅔ indent level? Nonsense!
Let me use analogy (I do know that analogies arent good arguments): the difference between using tabs or spaces is like difference between LaTeX and Word. Using tabs or LaTeX is like caring about structure first and then about form and do the opposite if you use spaces or Word.

For example, I want to put emphasis on something in my document.
LaTeX way of thinking: "I  use \emph{something} and don't care much, will it mean to write using bold font or red color".
Word way: "I will make this text red. It means emphasizing. Why? I like red." (yes, Word has styles, but who uses them?)
The same for tabs vs spaces. I want to indent code.
Tabs way: "I put 1 tab. This means and will always mean one indent level, no matter will it shift to 2nd, 3rd, 4th or 8th column in other users' editors."
Spaces way: "I will put N spaces. It will mean indentation. Why not M? Dunno. Maybe coz I've seen N red cars today and I like red. Ah, and I don't care if others prefer shifts to Mth column per indentation level."

Don't get me wrong. I don't say that form doesn't matter (after all, which "structure" could tabs show in curly-braced language?). Both tabs and spaces are used to make code readable. I speak about accents.